src/Umbraco.Web/Search/ExamineFinalComponent.cs

```diff
@@ -16,27 +18,65 @@ public sealed class ExamineFinalComponent : IComponent
         private readonly IExamineManager _examineManager;
         BackgroundIndexRebuilder _indexRebuilder;
         private readonly IMainDom _mainDom;
-        
-        public ExamineFinalComponent(IProfilingLogger logger, IExamineManager examineManager, BackgroundIndexRebuilder indexRebuilder, IMainDom mainDom)
+        private readonly ISyncBootStateAccessor _syncBootStateAccessor;
+        private readonly object _locker = new object();
+        private bool _initialized = false;
+
+        public ExamineFinalComponent(IProfilingLogger logger, IExamineManager examineManager, BackgroundIndexRebuilder indexRebuilder, IMainDom mainDom, ISyncBootStateAccessor syncBootStateAccessor)
         {
             _logger = logger;
             _examineManager = examineManager;
             _indexRebuilder = indexRebuilder;
             _mainDom = mainDom;
+            _syncBootStateAccessor = syncBootStateAccessor;
+
+            // must add the handler in the ctor because it will be too late in Initialize
+            // TODO: All of this boot synchronization for cold boot logic needs should be fixed in netcore
+            _syncBootStateAccessor.Booting += _syncBootStateAccessor_Booting;
         }
 
-        public void Initialize()
+        /// <summary>
+        /// Once the boot state is known we can see if we require rebuilds
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void _syncBootStateAccessor_Booting(object sender, SyncBootState e)
+        {
+            UmbracoModule.RouteAttempt += UmbracoModule_RouteAttempt;
+        }
+
+        private void UmbracoModule_RouteAttempt(object sender, RoutableAttemptEventArgs e)
         {
-            if (!_mainDom.IsMainDom) return;
+            UmbracoModule.RouteAttempt -= UmbracoModule_RouteAttempt;
 
-            _examineManager.ConfigureIndexes(_mainDom, _logger);
+            if (!_initialized)
+            {
+                lock (_locker)
+                {
+                    // double check lock, we must only do this once
+                    if (!_initialized)
+                    {
+                        if (!_mainDom.IsMainDom) return;
 
-            // TODO: Instead of waiting 5000 ms, we could add an event handler on to fulfilling the first request, then start?
-            _indexRebuilder.RebuildIndexes(true, 5000);
+                        var bootState = _syncBootStateAccessor.GetSyncBootState();
+
+                        _examineManager.ConfigureIndexes(_mainDom, _logger);
+
+                        // if it's a cold boot, rebuild all indexes including non-empty ones
+                        _indexRebuilder.RebuildIndexes(bootState != SyncBootState.ColdBoot, 0);
```

Comment:
Perhaps some delay may still be appropriate as the first few requests may be expensive till caches are built and indexing is very cpu/memory expensive.

Reply:
yep sure, we do that in netcore now too